### PR TITLE
[libc] Move printf long double to simple calc

### PR DESCRIPTION
**Pull request by michaelrj-google**
The Ryu algorithm is very fast with its table, but that table grows too
large for long doubles. This patch adds a method of calculating the
digits of long doubles using just wide integers and fast modulo
operations. This results in significant performance improvements vs the
previous int calc mode, while taking up a similar amound of peak memory.
It will be slow in some %e/%g cases, but reasonable fast for %f with no
loss of accuracy.

### CODE REVIEW

File: libc/docs/dev/printf_behavior.rst [hunk 1]
```diff
@@ -87,14 +87,25 @@ are not recommended to be adjusted except by persons familiar with the Printf
 Ryu Algorithm. Additionally they have no effect when float conversions are
 disabled.
 
+LIBC_COPT_FLOAT_TO_STR_NO_SPECIALIZE_LD
+---------------------------------------
+This flag disables the separate long double conversion implementation. It is
+not based on the Ryu algorithm, instead generating the digits by
+multiplying/dividing the written-out number by 10^9 to get blocks. It's
+significantly faster than INT_CALC, only about 10x slower than MEGA_TABLE,
```
Comment (gchatelet) [inline libc/docs/dev/printf_behavior.rst:95]: Is this 10% or 10 times? 10x seems like a lot.

File: libc/src/__support/float_to_string.h [hunk 2]
```diff
@@ -13,18 +13,27 @@
...
 // This file has 5 compile-time flags to allow the user to configure the float
-// to string behavior. These allow the user to select which 2 of the 3 useful
-// properties they want. The useful properties are:
-//  1) Speed of Evaluation
-//  2) Small Size of Binary
-//  3) Centered Output Value
-// These are explained below with the flags that are missing each one.
+// to string behavior. These were used to explore tradeoffs during the design
+// phase, and can still be used to gain specific properties. Unless you
+// specifically know what you're doing, you should leave all these flags off.
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:26]: > you should leave all these flags off

Isn't `LIBC_COPT_FLOAT_TO_STR_NO_SPECIALIZE_LD` true by default?

File: libc/src/__support/float_to_string.h [hunk 3]
```diff
@@ -582,182 +590,259 @@ class FloatToString {
...
-  if (exponent >= -FRACTION_LEN) {
-    const uint32_t idx =
-        exponent < 0
-            ? 0
-            : static_cast<uint32_t>(exponent + (IDX_SIZE - 1)) / IDX_SIZE;
-    const uint32_t len = internal::length_for_num(idx * IDX_SIZE, FRACTION_LEN);
-    return len;
-  } else {
-    return 0;
+template <> class FloatToString<long double> {
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:634]: Not in this patch but ideally `FloatToString` should be templated with the `FPType` enum and then use `FloatToString<get_fp_type<T>()>` to instantiate it. This way you can specialize the algorithm using `FPType::X86_Binary80` which is more precise than `long double`.

File: libc/src/__support/float_to_string.h [hunk 4]
```diff
@@ -582,182 +590,259 @@ class FloatToString {
...
-  return 0;
-  // TODO (michaelrj): Find a good algorithm for this that doesn't use a table.
-#endif
-}
+  LIBC_INLINE static constexpr void zero_leading_digits(
+      cpp::BigInt<FLOAT_AS_INT_WIDTH + EXTRA_INT_WIDTH, false> &int_num) {
+    // 64 is the width of the numbers used to internally represent the BigInt
+    for (size_t i = 0; i < EXTRA_INT_WIDTH / 64; ++i) {
+      int_num[i + (FLOAT_AS_INT_WIDTH / 64)] = 0;
+    }
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:678]: This feels like abstraction leakage.

File: libc/src/__support/float_to_string.h [hunk 5]
```diff
@@ -582,182 +590,259 @@ class FloatToString {
...
-FloatToString<long double>::zero_blocks_after_point() {
-#ifdef LIBC_COPT_FLOAT_TO_STR_USE_MEGA_LONG_DOUBLE_TABLE
-  return MIN_BLOCK_2[-exponent / IDX_SIZE];
-#else
-  return 0;
-  // TODO (michaelrj): Find a good algorithm for this that doesn't use a table.
-#endif
-}
+  LIBC_INLINE static constexpr void zero_leading_digits(
+      cpp::BigInt<FLOAT_AS_INT_WIDTH + EXTRA_INT_WIDTH, false> &int_num) {
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:674]: `cpp::UInt<FLOAT_AS_INT_WIDTH + EXTRA_INT_WIDTH>` is clearer than `cpp::BigInt<FLOAT_AS_INT_WIDTH + EXTRA_INT_WIDTH, false>`

File: libc/src/__support/float_to_string.h [hunk 6]
```diff
@@ -582,182 +590,259 @@ class FloatToString {
...
-      val = internal::get_table_positive<2048>(pos_exp, block_index);
-    } else if (MAX_INT_SIZE < 4096) {
-      val = internal::get_table_positive<4096>(pos_exp, block_index);
-    } else if (MAX_INT_SIZE < 8192) {
-      val = internal::get_table_positive<8192>(pos_exp, block_index);
-    } else if (MAX_INT_SIZE < 16384) {
-      val = internal::get_table_positive<16384>(pos_exp, block_index);
     } else {
-      val = internal::get_table_positive<16384 + 128>(pos_exp, block_index);
+      // if the exponent not positive, then the number is at least partially
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:706]: if the exponent **is** not positive

File: libc/src/__support/float_to_string.h [hunk 7]
```diff
@@ -582,182 +590,259 @@ class FloatToString {
...
+      // below the decimal point. Shift left to make the int a fixed point
+      // representation with the decimal point after the top EXTRA_INT_WIDTH
+      // bits.
+      const int SHIFT_AMOUNT = FLOAT_AS_INT_WIDTH + exponent;
+      static_assert(EXTRA_INT_WIDTH >= sizeof(long double) * 8);
+      float_as_int.shift_left(SHIFT_AMOUNT);
+
+      // If there are still digits above the decimal point, handle those.
+      if (float_as_int.clz() < EXTRA_INT_WIDTH) {
+        cpp::BigInt<EXTRA_INT_WIDTH, false> above_decimal_point =
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:716]: ditto `UInt` vs `BigInt`

File: libc/src/__support/float_to_string.h [hunk 8]
```diff
@@ -582,182 +590,259 @@ class FloatToString {
...
-          internal::get_table_negative<16384>(idx * IDX_SIZE, block_index + 1);
+
+    // negative_block_index starts at 0 with the first block after the decimal
+    // point, and 1 with the second and so on. This converts to the same
+    // block_index used everywhere else.
+
+    int block_index = -1 - negative_block_index;
+
+    // If we're currently after the requested block (remember these are
+    // negative indices) the reset the number to the start. This is only
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:813]: > the reset the number

we reset?

**Review by gchatelet — COMMENTED** [review]
I find the code in `libc/src/__support/float_to_string.h` quite complicated. It's hard to grasp the overall algorithm, especially with the negative block indices. Is there a way to make it simpler? Maybe by introducing dedicated structures?

File: libc/docs/dev/printf_behavior.rst (hunk 1 above)
Comment (michaelrj-google) [reply libc/docs/dev/printf_behavior.rst:95]: 10 times. It a lot, but only sort of. I separated out the decimal (%f) long double tests and ran them both with the new specialization and with the mega table with `-O3` in release mode. On my workstation the mega table version took ~400 microseconds and the new version took ~8000 microseconds.

The reason I think this is a worthwhile tradeoff is because the mega table takes up a lot of space. Just the converter (found in `build/libc/src/stdio/printf_core/CMakeFiles/libc.src.stdio.printf_core.converter.dir/`) takes up  ~5 megabytes with the mega table, but only ~150 kilobytes with the specialization.

The other option I was considering for saving space was using integer calculation (`INT_CALC`) which takes up more space (~200 Kb) and time (~7,500,000 microseconds) so in comparison the new version looks quite good.

File: libc/src/__support/float_to_string.h (hunk 2 above)
Comment (michaelrj-google) [reply libc/src/__support/float_to_string.h:26]: none of the flags are set by default. Setting that flag turns off the new specialization.

File: libc/src/__support/float_to_string.h (hunk 3 above)
Comment (michaelrj-google) [reply libc/src/__support/float_to_string.h:634]: this algorithm will work exactly the same for 80 bit and 128 bit long doubles. They have the same max exponent, but even if they didn't the constants should be calculated to be correct for any provided floating point type.

File: libc/src/__support/float_to_string.h (hunk 4 above)
Comment (michaelrj-google) [reply libc/src/__support/float_to_string.h:675]: probably, but I don't see another way to efficiently zero these blocks. If this were a normally sized integer I'd either do a shift left then shift right or use a mask, but both of those would have significant overhead for this very large integer.

**Review by michaelrj-google — COMMENTED** [review]
I see what you mean about this being a rather complex patch, I've addressed the simple comments and will update this patch again when I have an easier-to-read design.

File: libc/docs/dev/printf_behavior.rst (hunk 1 above)
Comment (gchatelet) [reply libc/docs/dev/printf_behavior.rst:95]: Cache wise the 5MiB table seems pretty bad for a function that can be called frequently on multiple cores.

The sheer size is an issue in itself but cache invalidation is also a concern, libc instruction and data footprint should be small to allow for more application state to stay in the cache. How is the table accessed by the way? Are the indices about always the same or is it completely random? Is there a way to compress the table? Can we come up with a smaller table and interpolate between values without losing precision?

File: libc/src/__support/float_to_string.h (hunk 4 above)
Comment (gchatelet) [reply libc/src/__support/float_to_string.h:675]: We can derive 64 from the implementation instead of hardcoding it.
```
using element_type = cpp::remove_cvref_t<decltype(int_num[0])>;
constexpr size_t ELEMENT_WIDTH = sizeof(element_type) * CHAR_T;
```

Or change BigInt to expose this information,

File: libc/src/__support/float_to_string.h [hunk 9]
```diff
@@ -152,15 +157,17 @@ LIBC_INLINE constexpr uint32_t ceil_log10_pow2(const uint32_t e) {
   return log10_pow2(e) + 1;
 }
 
+LIBC_INLINE constexpr uint32_t div_ceil(const uint32_t num,
+                                        const uint32_t denom) {
+  return (num + (denom - 1)) / denom;
```
Comment (lntue) [inline libc/src/__support/float_to_string.h:162]: Can you double check if  this division is optimized to multiplication or shifts in the generated codes since `BLOCK_SIZE` is constant?

File: libc/docs/dev/printf_behavior.rst (hunk 1 above)
Comment (michaelrj-google) [reply libc/docs/dev/printf_behavior.rst:95]: The table is accessed sequentially. It's a flattened two-dimensional table, where the `POW10_OFFSET` array gives you the starting indexes of each value in the larger array `POW10_SPLIT`. The negative indices have a similar but slightly different layout. Each number will map to one of the `POW10_OFFSET` values, then have an offset from that into the larger array which will be incremented or decremented (depending on if this is the positive or negative exponent table).

The 5MB table is already compressed. If you look at the top of `ryu_long_double_constants.h` you'll see `TABLE_SHIFT_CONST`, `IDX_SIZE`, and `MID_INT_SIZE`. These are the constants we can adjust to try to shrink the table. There's a more comprehensive explanation of what these mean in `utils/mathtools/ryu_tablegen.py` but here's the short version:

`MID_INT_SIZE` is the size of each entry in the `POW10_SPLIT` array. This needs to be at least `TABLE_SHIFT_CONST + IDX_SIZE + sizeof(BLOCK_INT)` so that it can actually fit the values.

`TABLE_SHIFT_CONST` (called `CONSTANT` in `ryu_tablegen.py`) adjusts the precision of each entry in the array, with higher being more precise. It's 120 here because anything lower tends to cause errors to get into the result.

`IDX_SIZE` is the compression factor. It's the number of exponents that can be mapped onto one entry in `POW10_OFFSET`. From my testing it only works when it's a power of 2. Pushing this higher would require increasing `MID_INT_SIZE` a lot, which would significantly reduce the actual size savings, and would also make the calculations slower.

I talked a bit with Ulf Adams (the original designer of the Ryu algorithm) and he suggested that it might be possible to also compress the table by approximating the next value by multiplying by `5**9`. In my testing this worked, but required a higher `TABLE_SHIFT_CONST` since you lose some precision with each approximation. In the end this only shrank the table a bit, and so I decided it wasn't worthwhile.

In conclusion, I believe the table is already compressed within an order of magnitude of its minimum size, and that makes it too large to be practical. Long doubles are rarely used, so carrying a large table all the time to speed them up seems like a bad idea.

File: libc/src/__support/float_to_string.h (hunk 9 above)
Comment (michaelrj-google) [reply libc/src/__support/float_to_string.h:162]: It definitely is, look at this godbolt link: https://godbolt.org/z/czh87M61b

File: libc/src/__support/float_to_string.h (hunk 4 above)
Comment (michaelrj-google) [reply libc/src/__support/float_to_string.h:675]: I've added a `WORD_SIZE` static member to `BigInt`.

**Review by michaelrj-google — COMMENTED** [review]
I've addressed more comments, and am still thinking about if there's a cleaner and more readable design.

**Comment by michaelrj-google** [comment]
I've updated the design to hopefully make it easier to follow.

File: libc/src/__support/float_to_string.h [hunk 10]
```diff
@@ -582,182 +590,269 @@ class FloatToString {
...
+  }
+
+  LIBC_INLINE constexpr BlockInt get_block(int block_index) {
+    if (block_index >= 0) {
+      return get_positive_block(block_index);
     } else {
-      val = internal::get_table_negative<16384 + 8192>(idx * IDX_SIZE,
-                                                       block_index + 1);
+      return get_negative_block(-1 - block_index);
     }
```
Comment (nickdesaulniers) [inline libc/src/__support/float_to_string.h:850]: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

Comment (nickdesaulniers) [reply libc/src/__support/float_to_string.h:850]: also, the `else` can be removed.

```c
if (foo)
   return ...
return ...
```

**Review by nickdesaulniers — COMMENTED** [review]
drive by style comments, feel free to address or not.

File: libc/src/__support/float_to_string.h [hunk 11]
```diff
@@ -582,182 +589,254 @@ class FloatToString {
...
+    // The decimal representation of 2**(-i) will have exactly i digits after
+    // the decimal point.
+    int num_requested_digits =
+        static_cast<int>((negative_block_index + 1) * BLOCK_SIZE);
 
-    cpp::UInt<MID_INT_SIZE> val;
-#ifdef LIBC_COPT_FLOAT_TO_STR_USE_MEGA_LONG_DOUBLE_TABLE
-    // ------------------------------ TABLE MODE -------------------------------
-    const int32_t SHIFT_CONST = TABLE_SHIFT_CONST;
+    return num_requested_digits > -exponent;
```
Comment (nickdesaulniers) [inline libc/src/__support/float_to_string.h:783]: This code is duplicated above (grep `num_requested_digits`).  Is there a way to better share/DRY up this code?

File: libc/src/__support/float_to_string.h [hunk 12]
```diff
@@ -582,182 +589,254 @@ class FloatToString {
...
 #ifdef LIBC_COPT_FLOAT_TO_STR_NO_TABLE
+    if (exponent < -FRACTION_LEN) {
+      const int pos_exp = -exponent - 1;
+      const uint32_t pos_idx =
+          static_cast<uint32_t>(pos_exp + (IDX_SIZE - 1)) / IDX_SIZE;
+      const int32_t pos_len = ((internal::ceil_log10_pow2(pos_idx * IDX_SIZE) -
+                                internal::ceil_log10_pow2(FRACTION_LEN + 1)) /
+                               BLOCK_SIZE) -
+                              1;
+      return len = static_cast<uint32_t>(pos_len > 0 ? pos_len : 0);
```
Comment (nickdesaulniers) [inline libc/src/__support/float_to_string.h:619]: was assignment intentional? if not, please remove.

File: libc/src/__support/float_to_string.h [hunk 13]
```diff
@@ -582,182 +589,254 @@ class FloatToString {
...
-#else
-  return 0;
-  // TODO (michaelrj): Find a good algorithm for this that doesn't use a table.
-#endif
-}
+  LIBC_INLINE static constexpr void zero_leading_digits(
+      cpp::UInt<FLOAT_AS_INT_WIDTH + EXTRA_INT_WIDTH> &int_num) {
+    // WORD_SIZE is the width of the numbers used to internally represent the
+    // UInt
+    for (size_t i = 0; i < EXTRA_INT_WIDTH / int_num.WORD_SIZE; ++i) {
```
Comment (nickdesaulniers) [inline libc/src/__support/float_to_string.h:673]: Consider making this:

```c
for (size_t i = 0, e = EXTRA_INT_WIDTH / int_num.WORD_SIZE; i != e; ++i)
```

File: libc/src/__support/float_to_string.h [hunk 14]
```diff
@@ -582,182 +589,254 @@ class FloatToString {
...
+      int_num[i + (FLOAT_AS_INT_WIDTH / int_num.WORD_SIZE)] = 0;
+    }
```
Comment (nickdesaulniers) [inline libc/src/__support/float_to_string.h:675]: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

File: libc/src/__support/float_to_string.h [hunk 15]
```diff
@@ -582,182 +589,254 @@ class FloatToString {
...
+
+  LIBC_INLINE constexpr BlockInt get_negative_block(int negative_block_index) {
+    if (exponent >= 0)
+      return 0;
+
+    // negative_block_index starts at 0 with the first block after the decimal
+    // point, and 1 with the second and so on. This converts to the same
+    // block_index used everywhere else.
+
+    int block_index = -1 - negative_block_index;
```
Comment (nickdesaulniers) [inline libc/src/__support/float_to_string.h:803]: const

Comment (nickdesaulniers) [reply libc/src/__support/float_to_string.h:803]: Consider any other newly introduced variables if they should be const as well.  You're pretty good about marking things constexpr or const (I'm generally happy to have that qualifier used).  There was perhaps a few places where the qualifiers are missing and look like they could be added.

File: libc/src/__support/float_to_string.h (hunk 13 above)
Comment (michaelrj-google) [reply libc/src/__support/float_to_string.h]: `EXTRA_INT_WIDTH` and `int_num.WORD_SIZE` are both constexpr, so putting `EXTRA_INT_WIDTH / int_num.WORD_SIZE` into a variable is unnecessary.

File: libc/src/__support/float_to_string.h (hunk 14 above)
Comment (michaelrj-google) [reply libc/src/__support/float_to_string.h]: you can just say that the braces are unnecessary. You don't need to link to the style guide every time.

File: libc/src/__support/float_to_string.h (hunk 11 above)
Comment (michaelrj-google) [reply libc/src/__support/float_to_string.h]: This needs to be a template specialization so that I can add the extra class members it needs, so I also have to redefine these functions even though they're the same.

**Review by nickdesaulniers — APPROVED** [review]

**Comment by michaelrj-google** [comment]
@gchatelet and @lntue do you have any comments before I land this? I'm going to let it run with the fuzzer for a while to check correctness.

File: libc/src/__support/UInt.h [hunk 16]
```diff
@@ -27,6 +27,11 @@ namespace LIBC_NAMESPACE::cpp {
 
 template <size_t Bits, bool Signed> struct BigInt {
 
+  // This being hardcoded as 64 is okay because we're using uint64_t as our
```
Comment (gchatelet) [inline libc/src/__support/UInt.h:30]: Can we also add
```
using word_type = uint64_t;
LIBC_INLINE_VAR static constexpr size_t WORD_SIZE = sizeof(word_type) * CHAR_T;
...
cpp::array<word_type, WORDCOUNT> val{};
```

File: libc/test/src/stdio/sprintf_test.cpp [hunk 17]
```diff
@@ -3072,6 +2972,157 @@ TEST_F(LlvmLibcSPrintfTest, FloatAutoConv) {
...
+      "462337569019001580859906425911782967213265389744605395555069797947978230"
+      "708108432086217134763779632408473684293543722127232658767439906910370146"
+      "716836295909075482355827087389127370874842532825987593970846704144140471"
+      "956027276735614286138656432085771988513977140957180090146798065497158947"
+      "229765733489703157617307078835099906185890777007500964162371428641176460"
+      "739074789794941408428328217107759915202650066155868439585510978709442590"
+      "231934194956788626761834746430104077432547436359522462253411168467463134"
+      "24896.000000");
+*/
+  /*
```
Comment (gchatelet) [inline libc/test/src/stdio/sprintf_test.cpp:3096]: same here

File: libc/src/__support/float_to_string.h [hunk 18]
```diff
@@ -567,197 +574,264 @@ class FloatToString {
...
+  static constexpr int FRACTION_LEN = fputil::FPBits<long double>::FRACTION_LEN;
+  static constexpr int EXP_BIAS = fputil::FPBits<long double>::EXP_BIAS;
+
+  static constexpr size_t FLOAT_AS_INT_WIDTH =
+      internal::div_ceil(fputil::FPBits<long double>::MAX_BIASED_EXPONENT -
+                             FPBits::EXP_BIAS,
+                         64) *
+      64;
+  static constexpr size_t EXTRA_INT_WIDTH =
+      internal::div_ceil(sizeof(long double) * 8, 64) * 64;
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:643]: nit: `CHAR_T` instead of `8`

File: libc/src/__support/float_to_string.h (hunk 3 above)
Comment (lntue) [reply libc/src/__support/float_to_string.h:628]: Can you add that point to a comment for this function, in case @gchatelet 's question comes back again in the future?

File: libc/src/__support/float_to_string.h [hunk 19]
```diff
@@ -567,197 +574,264 @@ class FloatToString {
...
+
+  // float_as_fixed represents the floating point number as a fixed point number
+  // with the point EXTRA_INT_WIDTH bits from the left of the number. This can
+  // store any number with a negative exponent.
+  cpp::UInt<FLOAT_AS_INT_WIDTH + EXTRA_INT_WIDTH> float_as_fixed = 0;
```
Comment (gchatelet) [inline libc/src/__support/float_to_string.h:648]: Can you define a type for `cpp::UInt<FLOAT_AS_INT_WIDTH + EXTRA_INT_WIDTH>`?
This way you can do `UintType::WORD_SIZE` instead of `int_num.WORD_SIZE` below.

File: libc/src/__support/float_to_string.h [hunk 20]
```diff
@@ -567,197 +574,264 @@ class FloatToString {
...
+  int int_block_index = 0;
+
+  static constexpr size_t BLOCK_BUFFER_LEN =
+      internal::div_ceil(internal::log10_pow2(FLOAT_AS_INT_WIDTH), BLOCK_SIZE);
+  BlockInt block_buffer[BLOCK_BUFFER_LEN] = {0};
+  size_t block_buffer_valid = 0;
+
+  template <size_t Bits>
+  LIBC_INLINE static constexpr BlockInt grab_digits(cpp::UInt<Bits> &int_num) {
+    auto wide_result = int_num.div_uint32_times_pow_2(1953125, 9);
```
Comment (lntue) [inline libc/src/__support/float_to_string.h:658]: This magic number is defined in other functions as `FIVE_EXP_NINE`.  Can you move it to a common definition at the beginning, and probably change to `EXP5_NINE` or `EXP5_9` similar to `exp2*` and `exp10`* functions.

File: libc/src/__support/float_to_string.h (hunk 9 above)
Comment (lntue) [reply libc/src/__support/float_to_string.h:161]: `const` quantifiers in the function signatures don't do anything with value types.  You should remove them to reduce confusion with the function signatures.  Same with the function below.

File: libc/src/__support/UInt.h (hunk 16 above)
Comment (michaelrj-google) [reply libc/src/__support/UInt.h:30]: I've done that, though I'm leaving the large scale refactoring for a followup patch.

**Comment by michaelrj-google** [comment]
Rebased, going to land once presubmits finish.
